### PR TITLE
feat(runner): provide test context to `test.each`

### PR DESCRIPTION
**Pull request by hi-ogawa**
### Description

- Closes https://github.com/vitest-dev/vitest/issues/4642

I made a few prototypes here in this PR based on the suggestions in the issue.

todo

- [ ] finalize api (Now this PR includes only `each7`)
  - last argument? [issuecomment-1871372078](https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1871372078)
    - see `test.each5` and `test.each7`
    - currently fixture extraction looks for first argument, so it needs to adjust `getUsedProps` slightly, but it should be simple.
    - the usage of spreading `(...args, { expect }) => ...` is not possible, but overall, this feels least breaking and least surprising in terms of API.
  - first argument? [issuecomment-1859974766](https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1859974766)
    - see `test.each6`
    - based on context's usage, the user needs to switch between `({ expect }, arg) => ...` and `(arg) => ...`, which might be un-intuitive than having it in the last argument. Also, if Vitest would change the default to be `{ context: true }` in the future, then all users (regardless of context/fixture usage) would be impacted by breaking change. Also this would break Jest compatibility.
  - inside test context object? [issuecomment-1894155135](https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1894155135) 
    - see `test.each3`
    - the downside would be similar to "first argument" explained above. But if ever, breaking change is possible, then this API feels best with Vitest's context concept.
  - where to `boolean` flag?
    - `test.each([...])("title", (...) => { ... }, { context: true })` [issuecomment-1894155135
](https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1894155135)
      - see `test.each7`
    - `test.each([...], { context: true })` [issuecomment-1859974766](https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1859974766)
      - see `test.each5`
      - it cannot support string literal usage `` test.each`${0} | ${1}` ``
- [x] ts function overload
- [ ] support `TestOption.context` at the 2nd position https://github.com/vitest-dev/vitest/pull/5142
- [x] test + type testing
- [x] docs

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: test/core/test/repro.test.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  async (...args) => {
+    await sleep(200)
+    expect(args[0]).toBeTypeOf('string')
+    expect(args[1]).toBeTypeOf('number')
+    args[0] satisfies string
+    args[1] satisfies number
+  },
+)
+
+myTest.concurrent.each6([['hello'], ['hi']], { context: true })(
```
Comment (jakebailey) [inline test/core/test/repro.test.ts:175]: I posted a writeup in https://github.com/vitest-dev/vitest/issues/4642#issuecomment-1898866866, but I suspect that `each6` here is the most compatible option. Note that you can also write `{ expect }, ...args` with this form, but you can't with `each5`.

Comment (sheremet-va) [reply test/core/test/repro.test.ts:175]: There is also another syntax for `each` using the template literal:

```ts
test.each`
  name | description 
  ${1} | ${"one"}
`("test", () => {})
```

So I am not sure we can have a second argument in each.

Comment (jakebailey) [reply test/core/test/repro.test.ts:175]: Oh, right, that yucky thing :(

File: packages/runner/src/types/tasks.ts [hunk 2]
```diff
@@ -192,7 +212,6 @@ export interface TestOptions {
 }
 
 interface ExtendedAPI<ExtraContext> {
-  each: TestEachFunction
```
Comment (hi-ogawa) [inline packages/runner/src/types/tasks.ts:195]: `each` in `ExtendedAPI` (here) and `SuiteAPI` (below) seems to be redundant.

**Comment by hi-ogawa** [comment]
Superseded by
- https://github.com/vitest-dev/vitest/pull/5861

**Pull request closed without merging**